### DISCUSSION TITLE
react-remove-scroll is shifting my fixed header every time a radix select is opened

### DISCUSSION THREAD

**Discussion by jtabone16** [radix-ui/primitives#1586] [Help] [answered]
I've seen https://github.com/radix-ui/primitives/discussions/1340 and https://github.com/radix-ui/primitives/discussions/1100 and updating radix select to 1.0.0 doesn't fix the issue. Seeing this on Mac and Windows on all browsers.  

Note that `overflow: hidden !important;` and `margin-right: 15px !important;` on the body via `react-remove-scroll` is causing the issue

**Comment by benoitgrelard** [comment, marked as answer]
Hi @jtabone16,

This is due to how `react-remove-scroll` handles the scrollbar removal technique.
When you have fixed elements, you have do add some additional classes to those elements to avoid them shifting.

We have to do it to our sidenav in our own docs for the same reason:

https://github.com/radix-ui/website/blob/8c5a605f07879131e0f7a7e3fd777bb3604672d1/pages/docs/design-system/overview/%5Bslug%5D.tsx#L34-L41

**Reply by benoitgrelard** [reply]
On another note, we have been working on a potential replacement for `react-remove-scroll` as part of a hack week project which wouldn't require those kinds of shenanigans.

**Reply by jtabone16** [reply]
Gotcha. So our fixed header is in Angular and we're rendering React components in Angular as we port everything to React. We're currently rendering a Radix select in one of those React components. How would you recommend I handle applying that class in Angular?

**Reply by benoitgrelard** [reply]
I don't know much about Angular these days, but `RemoveScroll.classNames.zeroRight` is literally a string for a given className as you can see here: https://github.com/theKashey/react-remove-scroll-bar/blob/master/src/constants.ts#L1 so whichever way you use to set a class in angular, you just import that package and use that constant.

**Reply by jtabone16** [reply]
Thanks! Couldn't find where that was being defined

**Reply by jtabone16** [reply]
Hmm now it's shifting on the left side

**Reply by benoitgrelard** [reply]
It depends what is your fixed element:
- Is it a fixed element positioned on the right? Then use `RemoveScroll.classNames.zeroRight`
- Is it a fixed element that is full width? Then use `RemoveScroll.classNames.fullWidth`

**Reply by jtabone16** [reply]
it is full width and using that class doesn't fix the issue :/ it still shifts on the right side

**Reply by benoitgrelard** [reply]
We use it on our docs and it definitely works so perhaps there's something wrong in how yours is setup?
You can take a look at our implementation/styles here: https://github.com/radix-ui/website/blob/8c5a605f07879131e0f7a7e3fd777bb3604672d1/components/Header.tsx#L31

You can see the fixed header on any component docs pages: https://www.radix-ui.com/docs/primitives/components/dropdown-menu

**Reply by seanssel** [reply]
I still notice quite a bit of layout shift on your docs pages. Easier to see in the video in fullscreen, but you can see a bunch of stuff slightly shifting.



https://github.com/radix-ui/primitives/assets/2467750/533c644d-1a8c-4741-a933-cee844f4cb3d

**Reply by mk-logbuy** [reply]
> Is it a fixed element positioned on the right? Then use RemoveScroll.classNames.zeroRight
Is it a fixed element that is full width? Then use RemoveScroll.classNames.fullWidth

@benoitgrelard Legend, tyvm!

**Comment by 58bits** [comment]
Hi All - I've just bumped into this  as well, and have also noticed that either `react-remove-scroll` or radix adds the following css var to the body element when the select component is triggered...

```
body {
	--removed-body-scroll-bar-size: 17px
}
```

Since that var is available, I'm assuming (safely?) that it could be used to create a custom class with additional padding added to fixed elements as required in order to prevent the shift?

In our case, we import our Radix controls via a uikit in a monorepo, and so would prefer not to import the `react-remove-scroll` component directly into the application. 

And so instead we created a utility class that we can apply to fixed elements - like a fixed header, as follows:

```css
.fix-scroll-layout-shift-right {
  padding-right: var(--removed-body-scroll-bar-size);
}

.fix-scroll-layout-shift-left {
  padding-left: var(--removed-body-scroll-bar-size);
}
```

Are we miles out? Or does this look okay?

**Reply by hedysnike** [reply]
how do i keep the radix from resizing thou

**Comment by firxworx** [comment]
@benoitgrelard this is obviously a recurring annoyance... 

Throwing out there $0.02:

I think that Radix should in-house the scrollbar offset vs. relying on a third-party deps for something like this, and then provide devs with options for greater control. 

---

In my case I set scrollbars to always appear on the body so there is no "jump" when navigating between pages (e.g. one long enough to have a scroll vs. one that doesn't have a scroll). This is especially noticeable for PC users with a fat scrollbar. It neatly solves jank with modals etc. as well (except when using certain libs... including radix)

This simple trick also solves jank caused by modals hiding scrollbars when a layout's full-width backgrounds don't "stretch" to under the scrollbar (e.g. nav, body, footer different colours) (when this trick is done with a supporting modal implementation that can be configured to not hide the scrollbar and just lock the scroll so there is no jank / jump).

Following through chains of issues and discussions is a bit of a pain and from Google this comes up again and again
e.g. https://github.com/radix-ui/primitives/discussions/1100 

IMHO this is a sign that more flexible support is a _requirement_ to be embraced :) 

As a headless library meant to provide devs with drop-in a11y improvements to their project, I think it is in line with my understanding of the project goals, and enables radix to be better integrated with apps across Dialog, menus, etc. 

I suppose the same sentiment for first-class support includes for multiple/stacked modals and auto-triggered modals (such as session timeout warnings, or very "desktoppy" UX enterprise/business apps that might pop over with connectivity interruptions or important real-time business interruptions -- where there can be >1 even from totally different causes).

I love radix and would love to be able to use it for everything!

**Comment by ThePiyushAggarwal** [comment]
Can someone please help me out? Why does Radix UI disables the scrolling when the dropdown is open?

**Reply by firxworx** [reply]
@ThePiyushAggarwal this threw me off too when I first started using radix-ui. 

**Answer**: radix-ui treats the default behaviour of a dropdown menu as a "modal".

**Good news:** you can disable this behaivour.

Check out the docs w/ jumplink to the right section and look for _modal_ in the table of available props:
https://www.radix-ui.com/primitives/docs/components/dropdown-menu#root

The docs might not be clear especially for those who are not familiar with the concept of modals in general.
If that's you check out https://www.w3.org/WAI/ARIA/apg/patterns/dialog-modal/ (among other references). Generally this means focus lock, and the page underneath shouldn't be scrollable or interactive (in a modal dialog implementation you often see an overlay used that makes this more intuitive).

Personally I don't think this default aligns with how a dropdown is generally understood or implemented in the majority of UI's I come across.

Either way, you have your answer my friend. 

Unfortunately per my comment above the use of third party libraries to achieve "modal" is pretty inflexible and obtrusive w/ limited configuration options. Even radix can't escape it and there's code to try to workaround in places re the docs! For example at the above docs link, toggle the dropdown menu in the preview and notice how the topmost horizontal nav "janks" and layout shifts.

**Reply by ThePiyushAggarwal** [reply]
Thanks for this! I could solve the issue.

**Comment by sannajammeh** [comment]
The obvious problem here is that this conflicts with now stable API `scrollbar-gutter: stable`.
<img width="234" alt="image" src="https://github.com/radix-ui/primitives/assets/50969683/da092315-d9ef-48f9-9229-8d855b86de78">

This is the culprit. Two problems here: 
1. it's added as a general style tag on the`head` so overriding becomes difficult
2. it's using `!important` completely removing our way to override.

### Fix when using `scrollbar-gutter`:

Postcss nested or Sass 
```css
html {
   scrollbar-gutter: stable;
}
/* html tag is required here to provide higher specificity */
html {
  @supports (scrollbar-gutter: stable) {
    body {
      margin: 0 !important;
    }
  }
}

```

**Reply by brianlovin** [reply]
This fix worked for us when using `scrollbar-gutter: stable` — thanks!

**Comment by bepyan** [comment]
In my case, simply removing the `width: 100%` from the `body` works perfectly fine.
Maybe use `100vw` instead.

```css
html,
body {
  width: 100%; /* <<< remove it */
  height: 100%;
}
```

**Comment by LeonLiu2020** [comment]
Any new solutions?
The elements inside my sticky header navbar shift because the scrollbar disappears.

**Comment by Godrules500** [comment]
I have spent a very long time on this and just found this. Is there any fix for this in sight or alternatives?

**Comment by recronin** [comment]
## Complete Solution: scrollbar-gutter + Dynamic Gutter Width Measurement

After extensive testing, here's a solution that handles **all scenarios** without using `modal={false}`:

### The Problem

1. **Content changes** → scrollbar appears/disappears → layout shift
2. **Modal opens** → Radix removes scrollbar → layout shift  
3. **scrollbar-gutter: stable** + custom `::-webkit-scrollbar` → gutter width ≠ custom width → compensation mismatch

### The Solution

**1. CSS - Use scrollbar-gutter with modal override:**

```css
html {
  scrollbar-gutter: stable;
}

/* When modal opens, remove gutter and compensate with padding */
html:has([data-scroll-locked]) {
  scrollbar-gutter: auto;
}

body[data-scroll-locked] {
  /* Use measured gutter width, NOT --removed-body-scroll-bar-size */
  padding-right: var(--scrollbar-gutter-width, 15px) !important;
  margin-right: 0 !important;
}
```

**2. JS - Measure actual gutter width on mount:**

```tsx
function getSystemScrollbarWidth(): number {
  const outer = document.createElement("div")
  outer.style.cssText = `
    visibility: hidden;
    overflow: scroll;
    position: absolute;
    top: -9999px;
    width: 100px;
    height: 100px;
  `
  document.body.appendChild(outer)
  
  const inner = document.createElement("div")
  inner.style.width = "100%"
  outer.appendChild(inner)
  
  const scrollbarWidth = outer.offsetWidth - inner.offsetWidth
  outer.remove()
  return scrollbarWidth
}

// On app mount:
useEffect(() => {
  const gutterWidth = getSystemScrollbarWidth()
  document.documentElement.style.setProperty(
    "--scrollbar-gutter-width", 
    `${gutterWidth}px`
  )
}, [])
```

### Why This Works

| State | scrollbar-gutter | padding-right | Result |
|-------|-----------------|---------------|--------|
| Normal | stable | 0 | Gutter reserves space, no shift on content change |
| Modal open | auto | measured gutter width | Gutter removed, padding compensates, no shift |

### Key Insight

The issue with using `--removed-body-scroll-bar-size` is that it can be `0` when:
- `scrollbar-gutter: stable` is active but no scrollbar is present
- The calculation `window.innerWidth - document.documentElement.clientWidth` returns 0 because clientWidth already accounts for the gutter

By **measuring the actual system scrollbar width** (which equals gutter width), we get accurate compensation regardless of the page state.

Hope this helps others dealing with this issue! 🎉
